File: src/generator/stop-generator.ts
import { transit_realtime } from "../proto/gtfs-rt.js";
import { FileSpec, OrchestratorParams } from "../models.js";
import { trip_index } from "../proto/trip-index.js";
import { gtfs_api } from "../proto/output.js";
import { Generator } from "./generator.js";

export class StopGenerator extends Generator {

    generate(
        feed: transit_realtime.FeedMessage, 
        index: trip_index.TripIndex,
        params: OrchestratorParams
    ): FileSpec[] {
        console.log("Starting stop generation");
        var out = <FileSpec[]>[];

        for (const stopId of Object.keys(index.tripsByStop)) {
            const tripIds = index.tripsByStop[stopId].tripId;
            if (tripIds == null) continue;
            
            const message = this._createMessage(
                feed,
                params,
                tripIds
            );
            
            out.push(<FileSpec>{
                key: `canberra/v1/stop/${stopId}/live.pb`,
                contents: gtfs_api.RealtimeEndpoint.encode(message).finish(),
                mimeType: Generator.mimeType
            });
        }

        return out;
    }

}

File: core/src/generator/stop_generator.rs
use crate::generator::base::Generator;
use crate::generator::generic::GenericGenerator;
use crate::libs::gtfs_realtime::FeedMessage;
use crate::libs::trip_index::TripIndex;
use crate::models::{FileSpec, OrchestratorParams};

pub struct StopGenerator;
impl StopGenerator {
    pub fn new() -> Self {
        StopGenerator {}
    }
}
impl Generator for StopGenerator {
    fn generate(&self, feed: &FeedMessage, index: &TripIndex, params: &OrchestratorParams) -> Vec<FileSpec> {
        println!("Generating stops");
        GenericGenerator {
            prefix: "stop",
            extract: |i| &i.trips_by_stop,
            is_stop_id: true
        }.generate(feed, index, params)
    }

    fn single(&self, feed: &FeedMessage, index: &TripIndex, params: &OrchestratorParams, id: &str) -> Vec<u8> {
        GenericGenerator {
            prefix: "stop",
            extract: |i| &i.trips_by_route,
            is_stop_id: true
        }.single(feed, index, params, id)
    }
}